design-proposals/dra.md

```diff
@@ -0,0 +1,529 @@
+Note: All the following sections will assume that DRA feture gate is enabled
+
+### Virt controller changes
+
+1. If devices are requested using DRA, virt controller needs to render the virt-launcher manifest such that 
+   pod.spec.resourceClaims and pod.spec.containers.resources.claim sections are filled out.
+1. virt-controller needs a mechanism to watch for virt-launcher pods, resourceclaims and resourceslices to populate the 
+   vmi.status.deviceStatus using the steps mentioned in above section that has all the attributes (for example the 
+   pciAddress for the gpu device)
+
+### Virt launcher changes
+
+1. For devices generated using DRA, virt-launcher needs to use the vmi.status.deviceStatus to generate the domxml
+   instead of environment variables as in the case of device-plugins
+
+# Alternate Designs
+
+## Alternative 1
+
+```
+type GPU struct {
+	// Name of the GPU device as exposed by a device plugin
+	Name              string       `json:"name"`
+	DeviceName        string       `json:"deviceName"`
+	VirtualGPUOptions *VGPUOptions `json:"virtualGPUOptions,omitempty"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag   string `json:"tag,omitempty"`
+	Claim string `json:"claim,omitempty"`
+}
+
+type HostDevice struct {
+	Name string `json:"name"`
+	// DeviceName is the resource name of the host device exposed by a device plugin
+	DeviceName string `json:"deviceName"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag string `json:"tag,omitempty"`
+	// If specified, the ResourceName of the host device will be provisioned using DRA driver . which will not require the deviceName field
+	//+optional
+	ResourceClaim *ResourceClaim `json:"resourceClaim,omitempty"`
+}
+
+// ResourceClaim represents a resource claim to be used by the virtual machine
+type ResourceClaim struct {
+	// Name is the name of the resource claim
+	Name string `json:"name"`
+	// Source represents the source of the resource claim
+	Source ResourceClaimSource `json:"source"`
+}
+
+// ResourceClaimSource represents the source of a resource claim
+type ResourceClaimSource struct {
+	// ResourceClaimName is the name of the resource claim
+	ResourceClaimName string `json:"resourceClaimName"`
+	// ResourceClaimTemplateName is the name of the resource claim template
+	//
+	// Exactly one of ResourceClaimName and ResourceClaimTemplateName must
+	// be set.
+	ResourceClaimTemplateName string `json:"resourceClaimTemplateName"`
+}
+
+```
+
+This design misses the use-case where more than one DRA device is specified in the claim template, as each
+device will have its own template in the API.
+
+This design also assumes that the deviceName will be provided in the ClaimParameters, which requires the DRA drivers
+to have a ClaimParameters.spec.deviceName in their spec.
```

Comment:
From this description you are distinguishing the `name` and the `deviceName` which also refer to my previous questions. Maybe it will be clearer to add the difference between the 2 at the beginning